modules/apps/object/object-service/src/test/java/com/liferay/object/model/impl/ObjectDefinitionImplTest.java

```diff
@@ -32,76 +34,124 @@ public class ObjectDefinitionImplTest {
-			"/commerce-returns/commerce-return-items", "CommerceReturnItem",
-			"CommerceReturn");
+	public void testGetRESTContextPathRootDescendantNode() {
+
+		// Modifiable custom object definition
+
+		_testGetRESTContextPath(
+			"/c/rootobjects/customobjects", true, "CustomObject", "RootObject",
+			false);
+
+		// Modifiable system object definition
+
+		_testGetRESTContextPath(
+			"/headless-builder/applications/endpoints", true, "APIEndpoint",
+			"APIApplication", true);
+		_testGetRESTContextPath(
+			"/commerce-returns/commerce-return-items", true,
+			"CommerceReturnItem", "CommerceReturn", true);
 	}
 
-	private void _testGetRESTContextPathOfModifiableSystemObject(
-		String expectedRESTContextPath, String objectDefinitionName,
-		String rootObjectDefinitionName) {
+	private void _testGetRESTContextPath(
+		String expectedRESTContextPath, boolean modifiable,
+		String objectDefinitionName, String rootObjectDefinitionName,
+		boolean system) {
 
 		ObjectDefinition objectDefinition = Mockito.spy(
 			new ObjectDefinitionImpl());
 
-		objectDefinition.setModifiable(true);
+		objectDefinition.setModifiable(modifiable);
 		objectDefinition.setName(objectDefinitionName);
-		objectDefinition.setSystem(true);
+		objectDefinition.setSystem(system);
+
+		if (!modifiable && system) {
+			try {
+				objectDefinition.getRESTContextPath();
+				Assert.fail();
+			}
+			catch (UnsupportedOperationException
+						unsupportedOperationException) {
+
+				Assert.assertNotNull(unsupportedOperationException);
+			}
+		}
+		else {
+			if (!system) {
+				objectDefinition.setPluralLabel(
+					TextFormatter.formatPlural(
+						StringUtil.lowerCaseFirstLetter(objectDefinitionName)));
+			}
 
-		ObjectDefinitionLocalService objectDefinitionLocalService =
-			Mockito.mock(ObjectDefinitionLocalService.class);
+			ObjectDefinitionLocalService objectDefinitionLocalService =
+				Mockito.mock(ObjectDefinitionLocalService.class);
 
-		ReflectionTestUtil.setFieldValue(
-			ObjectDefinitionLocalServiceUtil.class, "_serviceSnapshot",
-			new Snapshot<ObjectDefinitionLocalService>(
-				ObjectDefinitionLocalServiceUtil.class,
-				ObjectDefinitionLocalService.class) {
+			ReflectionTestUtil.setFieldValue(
+				ObjectDefinitionLocalServiceUtil.class, "_serviceSnapshot",
+				new Snapshot<ObjectDefinitionLocalService>(
+					ObjectDefinitionLocalServiceUtil.class,
+					ObjectDefinitionLocalService.class) {
 
-				@Override
-				public ObjectDefinitionLocalService get() {
-					return objectDefinitionLocalService;
-				}
+					@Override
+					public ObjectDefinitionLocalService get() {
+						return objectDefinitionLocalService;
+					}
 
-			});
+				});
 
-		if (rootObjectDefinitionName != null) {
-			ObjectDefinition rootObjectDefinition = new ObjectDefinitionImpl();
+			if (rootObjectDefinitionName != null) {
+				ObjectDefinition rootObjectDefinition =
+					new ObjectDefinitionImpl();
 
-			rootObjectDefinition.setModifiable(true);
```

Comment:
I removed setting modifiable and system properties for root object as we don't need them for testing this method